driver-async/src/main/com/mongodb/async/client/gridfs/GridFSDownloadStreamImpl.java

```diff
@@ -232,18 +229,19 @@ private boolean hasFileInfo() {
         return hasInfo;
     }
 
-    private <T> void chunkNotFound(final int startChunkIndex, final SingleResultCallback<T> callback) {
-        callback.onResult(null, new MongoGridFSException(format("Could not find file chunk for file_id: %s at chunk index %s.",
-                fileInfo.getId(), startChunkIndex)));
+    private MongoGridFSException chunkNotFound(final int chunkIndex) {
+        return new MongoGridFSException(format("Could not find file chunk for files_id: %s at chunk index %s.", fileInfo.getId(),
+                chunkIndex));
     }
 
-    private void getBufferFromChunk(final Document chunk, final int expectedChunkIndex, final SingleResultCallback<byte[]> callback) {
+    private ChunkDataOrError getBufferFromChunk(final Document chunk, final int expectedChunkIndex) {
+        ChunkDataOrError chunkOrError = new ChunkDataOrError();
         if (chunk == null || chunk.getInteger("n") != expectedChunkIndex) {
-            chunkNotFound(expectedChunkIndex, callback);
-            return;
+            chunkOrError.set(null, chunkNotFound(expectedChunkIndex));
```

Comment:
nit: Get rid of the set method and just construct a new immutable instance when you need it.